### DISCUSSION TITLE
Enumのメソッド

### DISCUSSION THREAD

**Discussion by bigriv** [ebinase/othello#23] [Ideas]
Enum値の比較はインスタンスが持つ機能というよりは第三者が行うイメージなので、staticの方が良いのかなと思いました。
あと、インスタンスに比較の機能を持たせるとヌルポの原因にもなり兼ねないです。
（Color.equals(a, b)と書くことになるので、英文的に読めなくなってしまうのは難点）

https://github.com/ebinase/othello/blob/21201b23e1fed5d297d44b0fb9fe1d8552cbf89d/packages/Models/Board/Color/Color.php#L73

**Comment by ebinase** [comment]
ご指摘ありがとうございます！

まず、前提としてphpにはEnumが言語レベルでは実装されておらず、2021年11月25日リリースのPHP8.1でついにEnumがリリースされます。
本プロジェクトでもリリース直後にEnumへの換装を進めて行く予定となっています！
(仕様等の詳細はこちらを御覧ください：https://qiita.com/rana_kualu/items/bdfa6c844125c1d0f4d4)

それに伴い、php版のEnumが導入された前提でご回答いたします。

> **Enum値の比較はインスタンスが持つ機能というよりは第三者が行うイメージなので、staticの方が良いのかなと思いました。**
> **（Color.equals(a, b)と書くことになるので、英文的に読めなくなってしまうのは難点）**

結論としては比較メソッドは作らず等価演算子(===)の利用で良いと考えています。

今までは「Enum風」オブジェクトを自作していたため何をもって「等しい」とするかの定義がそれぞれのオブジェクトでバラバラでした。(Colorであればblackとwhiteの同一性判定に内部プロパティの$colorCodeを利用)
そのため「何を持って2つのオブジェクトが等しいとするか」という知識をクラス内に隠蔽するためにequalsメソッドを生やしていましたが、Enumが実装されることにより同一性の定義がより確実で簡素なものとなり等価演算子で簡単に比較できるようになったためせっかくなので今後はこの方法で統一していきたいと思っています。

```php
// Enumでの実装
$color = Color::from('white');
$color === Color::white;         // true
```

> **あと、インスタンスに比較の機能を持たせるとヌルポの原因にもなり兼ねないです。**

こちらのご指摘についてはphpの動的型付けの性質上、Enum型でありながらnullポインタを指すということは起こり得ないと考えています。(値にnullが設定されていれば有無を言わさずphp側でnull型として認識するため、ヌルポが出る前に型の不一致によるコンパイルエラーが発生します。)
また、EnumでないオブジェクトでEnum風の処理をしている場合であっても、本プロジェクトでは「完全コンストラクタパターン」を採用し正しい値を持ったオブジェクトのみ存在できる(=不正な値を持ったオブジェクトを作ろうとするとインスタンス化時に例外を投げる)ようにしているため、Colorオブジェクトであれば確実にblackかwhiteの値を保持していることを保証しています。

**Reply by bigriv** [reply]
・比較メソッドは作らず等価演算子(===)の利用を考えている
・ヌルポが出る前に型の不一致によるコンパイルエラーが発生する
Javaで考えていたので、全く考慮していませんでした・・・。ご回答ありがとうございます。
